### PR TITLE
Redirect user to login if missing valid access token

### PR DESCRIPTION
**Pull request by albennett**
## Description
Redirect user to login if missing valid access token

## Checklist

-   [ ] My code follows the code style and guidelines of this project. 
-   [ ] I have added tests to cover my changes.
-   [ ] All new and existing tests passed.
-   [ ] I have tested different user stories. 
-   [ ] I have identified similar or related features and tested they are still working.
-   [ ] I have performed a self-review of my own code.
-   [ ] I have tested my changes/additions in the latest Chrome, Firefox, and Safari.
-   [ ] I have commented my code, particularly in hard-to-understand areas.

#### Staging Deployment

To visit the URL of the staging deployment, please click "Show all checks" at the bottom of this PR and click "details" next to `bufferbotbrains/cicd-buffer-publish-legacy`

### CODE REVIEW

File: helm/buffer-publish.yaml [hunk 1]
```diff
@@ -59,6 +59,16 @@ env:
     value: https://k8s-api-php71.buffer.com
   - name: ANALYZE_API_ADDR
     value: https://analyze-api.buffer.com
+  - name: CLIENT_ID
+    valueFrom:
+      secretKeyRef:
+        key: client-id
+        name: buffer-publish-keys
```
Comment (karelm) [inline helm/buffer-publish.yaml]: is this in the secrets collection?

File: packages/server/middlewares/verifyAccessToken.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+const rp = require('request-promise');
+
+function redirect (res, url) {
+  res.redirect(307, url);
+}
+
+function isRequestingApp(req) {
+  return req.method === 'GET' && req.accepts('text/html');
```
Comment (djfarrelly) [inline packages/server/middlewares/verifyAccessToken.js]: Ah to make sure it's not a fetch/xhr right I'm guessing?

File: packages/server/middlewares/verifyAccessToken.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+function isAccessTokenValid(req) {
+  if (!req.session.publish.accessToken) {
+    return true;
```
Comment (djfarrelly) [inline packages/server/middlewares/verifyAccessToken.js]: Should this be false @albennett if the token isn't there?

File: packages/server/middlewares/verifyAccessToken.js (hunk 2 above)
Comment (federicoweber) [reply packages/server/middlewares/verifyAccessToken.js]: That's right @djfarrelly, this basically constrains it to the first app load, as it's quite an expensive check to perform for all requests ;)

File: packages/server/middlewares/verifyAccessToken.js (hunk 3 above)
Comment (federicoweber) [reply packages/server/middlewares/verifyAccessToken.js]: Hey @djfarrelly, this is set to `true`, to limit it to only invalid access tokens. In this way, if we don't have one the middleware will just invoke `next`, and let the rest of the app logic kick in.

File: packages/server/middlewares/verifyAccessToken.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    .catch(r => false);
+}
+
+module.exports = async (req, res, next) => {
+  // As this is a quite expensive check to perform, and we want the redirect to happen
+  // before the app is rendered, we are processing it only on the first request.
+  if (isRequestingApp(req)) {
+    const isValid = await isAccessTokenValid(req);
+    if (!isValid) {
+      return redirect(res, 'https://login.buffer.com/login?redirect=https://publish.buffer.com?skipImpersonation=true');
```
Comment (federicoweber) [inline packages/server/middlewares/verifyAccessToken.js]: Hey @albennett, we need this cause there could be some accounts for which the login service will erroneously think they are in impersonation mode, and won't re-issue the token. That was the problem I war experiencing Friday with the account redirect 😉.

Comment (albennett) [reply packages/server/middlewares/verifyAccessToken.js]: Thanks for adding this @federicoweber! Glad you figured out the account redirect issue 🙌

File: helm/buffer-publish.yaml (hunk 1 above)
Comment (albennett) [reply helm/buffer-publish.yaml]: Hey @karelm great question. We checked that this was present in the secrets collection :)

**Pull request merged**